src/main/java/net/dries007/tfc/common/blocks/devices/BloomeryBlock.java

```diff
@@ -0,0 +1,285 @@
+{
+    public static final DirectionProperty FACING = BlockStateProperties.HORIZONTAL_FACING;
+    public static final BooleanProperty LIT = BlockStateProperties.LIT;
+    public static final BooleanProperty OPEN = BlockStateProperties.OPEN;
+
+    public static final VoxelShape OPEN_NORTH_SHAPE = Shapes.or(
+        box(0D, 15D, 0D,16D, 16D, 2D),
+        box(0D, 0D, 0D, 16D, 1D, 2D),
+        box(0D, 1D, 0D, 1D, 15D, 1D),
+        box(15D, 1D, 0D, 16D, 15D, 1D),
+        box(0D, 1D, 1D, 2D, 15D, 8D),
+        box(14D, 1D, 1D, 16D, 15D, 8D)
+    );
+    public static final VoxelShape OPEN_SOUTH_SHAPE = Shapes.or(
+        box(0D, 15D, 14D, 16D, 16D, 16D),
+        box(0D, 0D, 14D, 16D, 1D, 16D),
+        box(0D, 1D, 15D, 1D, 15D, 16D),
+        box(15D, 1D, 15D, 16D, 15D, 16D),
+        box(0D, 1D, 8D, 2D, 15D, 15D),
+        box(14D, 1D, 8D, 16D, 15D, 15D)
+    );
+    public static final VoxelShape OPEN_WEST_SHAPE = Shapes.or(
+        box(0D, 15D, 0D,2D, 16D, 16D),
+        box(0D, 0D, 0D, 2D, 1D, 16D),
+        box(0D, 1D, 0D, 1D, 15D, 1D),
+        box(0D, 1D, 15D, 1D, 15D, 16D),
+        box(1D, 1D, 0D, 8D, 15D, 2D),
+        box(1D, 1D, 14D, 8D, 15D, 16D)
+    );
+    public static final VoxelShape OPEN_EAST_SHAPE = Shapes.or(
+        box(14D, 15D, 0D, 16D, 16D, 16D),
+        box(14D, 0D, 0D, 16D, 1D, 16D),
+        box(15D, 1D, 0D, 16D, 15D, 1D),
+        box(15D, 1D, 15D, 16D, 15D, 16D),
+        box(8D, 1D, 0D, 15D, 15D, 2D),
+        box(8D, 1D, 14D, 15D, 15D, 16D)
+    );
+
+    public static final VoxelShape CLOSED_NORTH_SHAPE = box(0D, 0D, 0D, 16D, 16D, 2D);
+    public static final VoxelShape CLOSED_SOUTH_SHAPE = box(0D, 0D, 14D, 16D, 16D, 16D);
+    public static final VoxelShape CLOSED_WEST_SHAPE = box(0D, 0D, 0D, 2D, 16D, 16D);
+    public static final VoxelShape CLOSED_EAST_SHAPE = box(14D, 0D, 0D, 16D, 16D, 16D);
+
+    private static final MultiBlock BLOOMERY_CHIMNEY; // Helper for determining how high the chimney is
+    private static final MultiBlock[] BLOOMERY_BASE; // If one of those is true, bloomery is formed and can operate (has at least one chimney)
+    private static final MultiBlock GATE_Z, GATE_X; // Determines if the gate can stay in place
+    private static final Direction[] NORTH_SOUTH_DOWN = new Direction[]{Direction.NORTH, Direction.SOUTH, Direction.DOWN};
+    private static final Direction[] EAST_WEST_DOWN = new Direction[]{Direction.EAST, Direction.WEST, Direction.DOWN};
+
+    static
+    {
+        BiPredicate<LevelAccessor, BlockPos> stoneMatcher = (level, pos) -> level.getBlockState(pos).is(TFCTags.Blocks.BLOOMERY_INSULATION) && level.getBlockState(pos).isCollisionShapeFullBlock(level, pos);
+        Predicate<BlockState> insideChimney = state -> state.getBlock() == TFCBlocks.MOLTEN.get() || state.getMaterial().isReplaceable();
+        Predicate<BlockState> center = state -> state.is(TFCBlocks.MOLTEN.get()) || state.is(TFCBlocks.BLOOM.get()) || state.getMaterial().isReplaceable();
```

Comment:
```suggestion
        Predicate<BlockState> center = state -> state.is(TFCBlocks.MOLTEN.get()) || state.is(TFCBlocks.BLOOM.get()) || state.isAir();
```
Unless there's a specific reason why this is allowed, same for the above line.